### DISCUSSION TITLE
Discussion on FastAPI performance

### DISCUSSION THREAD

**Discussion by wu-clan** [fastapi-practices/fastapi_best_architecture#265] [General]
After reviewing the fastapi-0.108.0 source code

## I have the following points:

1. response_model:
    - When using pydantic-v1, don't use response_model in the interface parameters and don't specify that the interface returns a model; before the interface responds, fastapi will first do a lot of adaptation and validation of the v1 model, then serialize it according to the model, and finally parse the response through jsonable_encoder. and return.
    - When using pydantic-v2, it is recommended to use it, fastapi will give the response content directly to pydantic to process (including validation) and return
2. jsonable_encoder:
    - Additional performance overhead (data parsing)
3. response_class:
    - fastapi will use the official json to encode the response content by default.

## How to improve performance?

1. when using pydantic-v2, prioritize the use of response_model
2. don't use jsonable_encoder unless necessary
3. high performance response_class

## What we've done so far:

Use jsonable_encoder to parse the response in advance, and then have fastapi to handle the content.

## TODO:

Obviously, we need to change this. The reason we use a custom ResponseBase is to unify the return structure and simplify the serialization configuration;

- [x] using the response_model globally
- [x] Do not use jsonable_encoder within ResponseModel
- [x] for serialization configuration, use interface parameter settings, e.g.: response_model_exclude...
- [x] high-performance serializer（Pydantic serialization enforces binding internally, which only works on the final serialization）

## Prompt but not implemented

1. Do not use response_model
    - Because Pydantic serialization enforces binding internally, if the interface returns Schema, the serialization speed will be affected by pydantic. The solution is to use MsgSpecJSONResponse serialization and return directly.

**Comment by wu-clan** [comment]
Fastapi makes heavy use of anyio to handle synchronization functions, but there is a performance overhead with anyio, check out the：https://github.com/agronholm/anyio/discussions/543

**Comment by wu-clan** [comment]
Json serialization comparison：https://www.libhunt.com/compare-orjson-vs-msgspec

msgspec docs: https://jcristharif.com/msgspec/benchmarks.html

**Reply by downdawn** [reply]
It seems that someone has already submitted a pull request (PR).
[https://github.com/tiangolo/fastapi/pull/10631](https://github.com/tiangolo/fastapi/pull/10631)

**Comment by wu-clan** [comment]
If you want to avoid pydantic as much as possible and have full performance, you need to completely rewrite the Route class, which is destructive (not recommended).

**Comment by wu-clan** [comment]
Performance optimization: #382

**Reply by wu-clan** [reply]
Related: #324

**Comment by wu-clan** [comment]
related: https://github.com/pydantic/pydantic/discussions/6388
